File: workflow-sdk-node/src/handler/registry.ts
import type { Job } from '../client/types.js'

/** Context passed to handler functions. */
export interface JobContext {
  jobId: string
  instanceId: string
  jobType: string
  retriesRemaining: number
  variables: Record<string, unknown>
}

/** Result returned by a handler on success. */
export interface HandlerResult {
  variablesToSet?: Record<string, unknown>
}

/** A job handler function. Throw NonRetryableError to skip the retry budget. */
export type Handler = (ctx: JobContext) => Promise<HandlerResult>

/** Throw from a Handler to bypass the retry budget (retryable=false). */
export class NonRetryableError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'NonRetryableError'
  }
}

/**
 * Registry maps jobType strings to Handler functions (R-010).
 */
export class HandlerRegistry {
  private readonly handlers = new Map<string, Handler>()

  /** Register a handler for a jobType. */
  register(jobType: string, handler: Handler): void {
    if (!jobType) throw new Error('jobType must not be empty')
    this.handlers.set(jobType, handler)
  }

  /** Returns the handler for jobType, or undefined. */
  get(jobType: string): Handler | undefined {
    return this.handlers.get(jobType)
  }

  /** Returns all registered jobType strings. */
  jobTypes(): string[] {
    return [...this.handlers.keys()]
  }

  /** Build a JobContext from an engine Job. */
  static contextFrom(job: Job): JobContext {
    return {
      jobId: job.id,
      instanceId: job.instanceId,
      jobType: job.jobType,
      retriesRemaining: job.retriesRemaining,
      variables: (job.variables as Record<string, unknown>) ?? {},
    }
  }
}

File: workflow-sdk-go/handler/registry.go
// Package handler provides the job-type handler registry.
// Handlers are keyed by jobType string — never by Java delegate class path (R-010).
package handler

import (
	"context"
	"fmt"
	"sync"
)

// JobContext gives handlers typed access to the job's variables and metadata.
type JobContext struct {
	JobID            string
	InstanceID       string
	StepID           string
	JobType          string
	Attempt          int
	RetriesRemaining int
	Variables        map[string]any
}

// Result is what a handler returns on success.
type Result struct {
	// VariablesToSet are merged back into the instance variables on CompleteJob.
	VariablesToSet map[string]any
}

// Handler is the function signature that workers implement.
type Handler func(ctx context.Context, job JobContext) (Result, error)

// Registry maps jobType strings to Handler functions.
type Registry struct {
	mu       sync.RWMutex
	handlers map[string]Handler
}

// New creates an empty Registry.
func New() *Registry {
	return &Registry{handlers: make(map[string]Handler)}
}

// Register associates jobType with handler. Panics if jobType is empty.
func (r *Registry) Register(jobType string, h Handler) {
	if jobType == "" {
		panic("workflow-sdk-go: Register called with empty jobType")
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	r.handlers[jobType] = h
}

// Get returns the handler for jobType or an error if none is registered.
func (r *Registry) Get(jobType string) (Handler, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	h, ok := r.handlers[jobType]
	if !ok {
		return nil, fmt.Errorf("no handler registered for jobType %q", jobType)
	}
	return h, nil
}

// JobTypes returns all registered jobType strings.
func (r *Registry) JobTypes() []string {
	r.mu.RLock()
	defer r.mu.RUnlock()
	types := make([]string, 0, len(r.handlers))
	for k := range r.handlers {
		types = append(types, k)
	}
	return types
}
